.pre-commit-config.yaml

```diff
@@ -0,0 +1,60 @@
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.4.0
+    hooks:
+      - id: check-yaml
+      - id: check-toml
+      - id: check-json
+        # package*.json files should already adhere to conventional standards since npm manages them.
+        exclude: "^tsconfig.json$"  # ignore problems that originate from upstream
+      - id: end-of-file-fixer
+        # ignore problems in pytest snapshots and those that originate from upstream
+        exclude: "^typings/.*|tests/snapshots/.*"
+      - id: trailing-whitespace
+      - id: mixed-line-ending
+        args: ["--fix=lf"]
+      - id: check-added-large-files
+        # may conflict with large pictures (or big additions to package-lock.json)
+        args: [--maxkb=1000] # (default is 500 kB)
+  - repo: https://github.com/psf/black
+    rev: '23.1.0'
+    hooks:
+      # we'll let this hook modify the files since it will fail if any files get changed
+      - id: black
+        exclude: "^docs/tensorstore_demo/.*\\.py$"
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.0.1
+    hooks:
+      - id: mypy
+        # this hook's behavior doesn't respect toml config about `files` and `exclude`.
+        # we must re-specify using a regex pattern here.
+        files: "^sphinx_immaterial/.*\\.py$|^tests/.*\\.py$"
+        exclude: ".*/setup\\.py$"
+        additional_dependencies: [
+          pydantic, # needed for pydantic.mypy plugin (as respected from toml)
```

Comment:
The duplication of our dependencies here and for pylint is the biggest thing I have an issue with.

There seems to be a `system` language option for pre-commit where it doesn't create a new venv -- perhaps that would allow us to avoid this duplication?

Reply:
Interesting, I'll look into that `system` option...

Reply:
## system's tool versions used as is
So, using the system hook presents a problem which pre-commit is designed to solve by default:
- pre-commit hooks that use `system` env are ignoring the `rev` field which is required in the config. This means that any version of tool installed in the system env is used without validating that the tool's version matches the specified `rev` tag.

To summarize, the versions specified in both pre-commit config and dev-requirements.txt are even more required when using the system env.

> The duplication of our dependencies here and for pylint is the biggest thing I have an issue with.
>
>There seems to be a system language option for pre-commit where it doesn't create a new venv -- perhaps that would allow us to avoid this duplication?

Yes, using the system env does not mandate that `additional_dependencies` be specified in pre-commit config.

The `black` tool doesn't need any additional deps, but using the system env means we can't cache the `black` tool's install.

## Caching the pre-commit's hooks' venvs doesn't cache the system env's site_pkgs

With pylint and mypy hooks using pre-commit's isolated venvs, the specified `additional_dependencies` are cached as well. Consequently, this means the CI runner's cache (of mypy and pylint pre-commit venvs) could be out-of-date until the cache expires (which I think happens after a week) or until the pre-commit config changes.

If we want to use `actions/cache` to cache mypy and pylint tools' system install (which inherently need the theme's deps as well), then we should also use stricter version specs in the theme's requirements.txt. For this reason, it makes sense to use the non-cached system env for the mypy and pylint hooks and only cache the pre-commit venvs used for the other checks.

Remember, we've had problems in the past with "silent" updates to the black, so it  makes sense to cache a specific version of black via pre-commit hook's venv.

## hooks specific to pre-commit
I have found a way to use [pre-commit/pre-commit-hooks](https://github.com/pre-commit/pre-commit-hooks) to check for

- CRLFs
- trailing whitespace
- yaml/toml/json syntax errors
- EoF is a single blank line

without needing pre-commit installed, but this means the CI steps will go from 1 `pre-commit run --all-files` to a multitude of steps that does the same thing as pre-commit. Again, I don't want to maintain a generic script that does the same as pre-commit.

Reply:
I agree that it would make sense to pin versions of all the tools in any case.  We don't want any usage of actions/cache to affect which versions are used (currently does not, I believe).

Note that on Windows outside of WSL you can still install/use bash --- e.g. a copy is included with git.

But probably it would be simplest to write the lint script in Python rather than shell script in any case.

Reply:
It just seems redundant and more work to write a repo-specific script. I've seen it done in other repos and it isn't a set-and-forget option, especially if we pin theme deps to a specific version.

In writing the coverage steps, I thought it might be easier to run unit tests on various python versions (and deps' versions like sphinx) using [tox](https://tox.wiki/en/latest/). I'll look into that...

Reply:
Not loving tox for its configuration... I think that [nox](https://nox.thea.codes/en/stable/index.html) could be the compromise we're looking for.

Reply:
After exploring nox, I think it is the route that would be preferable in terms of writing our own script. It is so flexible that we can rewrite the entire CI in the noxfile.py (a config similar to pytest script) with the exception of the pypi deployments steps, So I'm wondering how much we want to put in the noxfile.py (`npm run <script>` can be performed as well).

I'll probably segregate the dev-requirements.txt using nested requirements.txt files for different nox sessions.
```
# in dev-requirements.txt

-r requirements/pylint.txt
-r requirements/mypy.txt
-r requirements/optional-cpp.txt
-r requirements/optional-json.txt
-r requirements/optional-keys.txt
```
This way I can keep the sessions' env cleaner/quicker rather than blindly installing everything. We could even use this in setup.py (or pyproject.toml) as well instead of re-specifying the optional dependencies.

Reply:
I haven't used nox before, but it looks good.  I have used tox, but have not been all that happy with it.

Reply:
I'm having fun with nox... Maybe too much fun 🤓.

Once I'm ready to commit, I want to figure out how to get nox to use the various [python versions from the github-hosted runners' tool cache](https://github.com/actions/setup-python/blob/main/docs/advanced-usage.md#hosted-tool-cache). That way, when building the dists and docs, the CI won't have to create a separate job for each version of python, just a job for each OS. This idea may result in long CI runs though.

BTW, I included the docs' builds in the coverage data (locally), and the total coverage spiked to 84-89% (from 55-57% for only the unit tests) depending on which versions of sphinx, builder, and python I'm using in the nox session. This doesn't require the package installed because the `coverage` cmd adds the invoking directory to python's `sys.path`. I'm not sure if _all_ the monkey patching is observed in the coverage data though.